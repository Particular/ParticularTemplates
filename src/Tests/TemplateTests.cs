using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using NUnit.Framework;
using Particular.Approvals;

[TestFixture]
public class TemplateTests : IDisposable
{
    static TemplateTests()
    {
        // Windows separator char becomes special character in a regex, need to escape it
        var separator = Path.DirectorySeparatorChar == '\\' ? @"\\" : Path.DirectorySeparatorChar.ToString();
        IgnorePathRegex = new Regex($@"^{separator}(bin|obj){separator}", RegexOptions.Compiled | RegexOptions.IgnoreCase);
    }

    [OneTimeSetUp]
    public async Task Setup()
    {
        await Uninstall(CreateTimeoutToken());
        await Install(CreateTimeoutToken());
    }

    static async Task Install(CancellationToken cancellationToken)
    {
        //dotnet new --install \nugets\ParticularTemplates.xxx.nupkg
        var nugets = Path.GetFullPath(Path.Combine(ProjectDirectory.ProjectPath, "..", "..", "nugets"));
        var directoryInfo = new DirectoryInfo(nugets);
        var latestNuget = directoryInfo.GetFiles("*.nupkg")
            .OrderByDescending(f => f.LastWriteTimeUtc)
            .First();

        await DotNetTemplatesHelper.Install(latestNuget.FullName, cancellationToken);
    }

    async Task Uninstall(CancellationToken cancellationToken)
    {
        // dotnet new --uninstall ParticularTemplates
        await DotNetTemplatesHelper.Uninstall("ParticularTemplates", cancellationToken);
    }

    public void Dispose()
    {
        //  Uninstall();
    }

    CancellationToken CreateTimeoutToken() => new CancellationTokenSource(60_000).Token;

    [Test]
    public async Task NServiceBusEndpoint()
    {
        await VerifyAndBuild("nsbendpoint", CreateTimeoutToken());
    }

    [Test]
    [TestCase("LearningTransport")]
    [TestCase("AzureServiceBus")]
    [TestCase("AzureStorageQueues")]
    [TestCase("SQS")]
    [TestCase("RabbitMQ")]
    [TestCase("SQL")]
    public async Task NServiceBusEndpointTransports(string transport)
    {
        var parameters = new Dictionary<string, string> { { "transport", transport } };
        if (transport == "MSMQ")
        {
            parameters.Add("framework", "net48");
        }
        await VerifyAndBuild("nsbendpoint", CreateTimeoutToken(), parameters);
    }

    [Test]
    [TestCase("ConsoleApp")]
    [TestCase("WindowsService")]
    [TestCase("Docker")]
    public async Task NServiceBusEndpointHosting(string hosting)
    {
        var parameters = new Dictionary<string, string> { { "hosting", hosting } };
        await VerifyAndBuild("nsbendpoint", CreateTimeoutToken(), parameters);
    }

    [Test]
    [TestCase("net7.0")]
    [TestCase("net6.0")]
    [TestCase("net48")]
    [TestCase("net472")]
    public async Task NServiceBusEndpointTargetFramework(string framework)
    {
        var parameters = new Dictionary<string, string> { { "framework", framework } };
        await VerifyAndBuild("nsbendpoint", CreateTimeoutToken(), parameters);
    }

    [Test]
    [TestCase("LearningPersistence")]
    [TestCase("SQL")]
    [TestCase("CosmosDB")]
    [TestCase("AzureTable")]
    [TestCase("RavenDB")]
    [TestCase("MongoDB")]
    [TestCase("DynamoDB")]
    [TestCase("NonDurable")]
    public async Task NServiceBusEndpointPersistence(string persistence)
    {
        var parameters = new Dictionary<string, string> { { "persistence", persistence } };
        await VerifyAndBuild("nsbendpoint", CreateTimeoutToken(), parameters);
    }

    [Test]
    public async Task Handler()
    {
        var targetDirectory = ProjectDirectory.GetSandboxPath();
        var token = CreateTimeoutToken();
        var messageClass = @"
using NServiceBus;

public class TestMessage : ICommand { }";

        await DotNetTemplatesHelper.Run("nsbendpoint", targetDirectory, parameters: null, token);
        // Required for item template bindings to work
        await DotNetTemplatesHelper.Restore(targetDirectory, token);

        await DotNetTemplatesHelper.Run("nsbhandler", targetDirectory, new() { { "messagetype", "TestMessage" } }, token);
        await File.WriteAllTextAsync(Path.Combine(targetDirectory, "Messages.cs"), messageClass, token);
        await DotNetTemplatesHelper.Build(targetDirectory, token);
        VerifyDirectory(targetDirectory);
    }

    [Test]
    public async Task Saga()
    {
        var targetDirectory = ProjectDirectory.GetSandboxPath();
        var token = CreateTimeoutToken();
        var messagesClass = @"
using NServiceBus;

public class OrderPlaced : IEvent
{
    public string CorrelationId { get; set; }
}

public class OrderBilled : IEvent
{
    public string CorrelationId { get; set; }
}";

        await DotNetTemplatesHelper.Run("nsbendpoint", targetDirectory, parameters: null, token);
        // Required for item template bindings to work
        await DotNetTemplatesHelper.Restore(targetDirectory, token);

        await DotNetTemplatesHelper.Run("nsbsaga", targetDirectory, new() { { "name", "ShippingPolicy" }, { "messagetype1", "OrderPlaced" }, { "messagetype2", "OrderBilled" } }, token);
        await File.WriteAllTextAsync(Path.Combine(targetDirectory, "Messages.cs"), messagesClass, token);
        await DotNetTemplatesHelper.Build(targetDirectory, token);
        VerifyDirectory(targetDirectory);
    }

    static async Task VerifyAndBuild(string templateName, CancellationToken cancellationToken, Dictionary<string, string> parameters = null)
    {
        var targetDirectory = ProjectDirectory.GetSandboxPath();
        await DotNetTemplatesHelper.Run(templateName, targetDirectory, parameters, cancellationToken);
        await DotNetTemplatesHelper.Build(targetDirectory, cancellationToken);
        VerifyDirectory(targetDirectory);
    }

    static void VerifyDirectory(string targetDirectory)
    {
        var fileText = new StringBuilder();

        foreach (var file in Directory.EnumerateFiles(targetDirectory, "*.*", SearchOption.AllDirectories).OrderBy(file => file, StringComparer.Ordinal))
        {
            var relativePath = file.Substring(targetDirectory.Length);
            if (IgnorePathRegex.IsMatch(relativePath))
            {
                continue;
            }

            const string hr = "---------------------------------------------------------------";

            var filename = Path.GetFileName(file);
            var contents = File.ReadAllText(file);

            if (Path.GetExtension(filename) == ".csproj")
            {
                contents = Regex.Replace(contents, @"Version=""\d+\.\d+\.\d+""", "Version=\"(VERSION)\"");
            }

            fileText.AppendLine(hr);
            fileText.AppendLine(filename);
            fileText.AppendLine(hr);
            fileText.Append(contents);
            fileText.AppendLine();
        }

        Approver.Verify(fileText.ToString(), scenario: ProjectDirectory.GetScenarioFromTestArguments());
    }

    static readonly Regex IgnorePathRegex;

}