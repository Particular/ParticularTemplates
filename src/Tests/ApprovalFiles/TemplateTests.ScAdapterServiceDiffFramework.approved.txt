Host.cs =>
using System;
using System.Threading.Tasks;
using NServiceBus;
using NServiceBus.Logging;
using ServiceControl.TransportAdapter;

namespace ScAdapterServiceDiffFramework
{
    class Host
    {
        // TODO: optionally choose a custom logging library
        // https://docs.particular.net/nservicebus/logging/#custom-logging
        // LogManager.Use<TheLoggingFactory>();
        static readonly ILog log = LogManager.GetLogger<Host>();

        ITransportAdapter adapter;

        // TODO: give the adapter an appropriate name
        public string AdapterName => "TransportAdapter.ScAdapterServiceDiffFramework";

        public async Task Start()
        {
            try
            {
                var adapterConfig = new TransportAdapterConfig<LearningTransport, LearningTransport>(AdapterName);

                adapterConfig.CustomizeEndpointTransport(t =>
                {
                    //TODO: Customize the endpoint-facing side of the adapter
                    //Use exactly the same settings as in regular endpoints
                });

                adapterConfig.CustomizeServiceControlTransport(t =>
                {
                    //TODO: Customize the ServiceControl-facing side of the adapter
                    //e.g. specify the same connection string as ServiceControl uses.
                });

                adapter = TransportAdapter.Create(adapterConfig);

                await adapter.Start();
            }
            catch (Exception exception)
            {
                FailFast("Failed to start", exception);
            }
        }

        public async Task Stop()
        {
            try
            {
                // TODO: perform any futher shutdown operations before or after stopping the adapter
                await adapter.Stop();
            }
            catch (Exception ex)
            {
                FailFast("Failed to stop correctly.", ex);
            }
        }

        void FailFast(string message, Exception exception)
        {
            try
            {
                log.Fatal(message, exception);

                // TODO: when using an external logging framework it is important to flush any pending entries prior to calling FailFast
                // https://docs.particular.net/nservicebus/hosting/critical-errors#when-to-override-the-default-critical-error-action
            }
            finally
            {
                Environment.FailFast(message, exception);
            }
        }
    }
}



Program.cs =>
using System;
using System.Linq;
using System.ServiceProcess;
using System.Threading.Tasks;

namespace ScAdapterServiceDiffFramework
{
    static class Program
    {
        // TODO: consider using C# 7.1 or later, which will allow
        // removal of this method, and renaming of MainAsync to Main
        public static void Main(string[] args) => MainAsync(args).GetAwaiter().GetResult();

        public async static Task MainAsync(string[] args)
        {
            var host = new Host();

            // pass this command line option to run as a windows service
            if (args.Contains("--run-as-service"))
            {
                using (var windowsService = new WindowsService(host))
                {
                    ServiceBase.Run(windowsService);
                    return;
                }
            }

            Console.Title = host.AdapterName;

            var tcs = new TaskCompletionSource<object>();
            Console.CancelKeyPress += (sender, e) => { e.Cancel = true; tcs.SetResult(null); };

            await host.Start();
            await Console.Out.WriteLineAsync("Press Ctrl+C to exit...");

            await tcs.Task;
            await host.Stop();
        }
    }
}



ScAdapterServiceDiffFramework.csproj =>
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net472</TargetFramework>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
  </PropertyGroup>

  <ItemGroup>
    <Reference Include="System.ServiceProcess" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="NServiceBus" Version="7.2.0" />
    <PackageReference Include="NServiceBus.Raw" Version="3.0.0" />
    <PackageReference Include="ServiceControl.TransportAdapter" Version="2.0.0" />
  </ItemGroup>

</Project>


WindowsService.cs =>
using System.ServiceProcess;

namespace ScAdapterServiceDiffFramework
{
    class WindowsService : ServiceBase
    {
        readonly Host host;

        public WindowsService(Host host) => this.host = host;

        protected override void OnStart(string[] args) => host.Start().GetAwaiter().GetResult();

        protected override void OnStop() => host.Stop().GetAwaiter().GetResult();
    }
}



